packages/core/src/saml-applications/routes/anonymous.ts

```diff
@@ -29,4 +49,65 @@ export default function samlApplicationAnonymousRoutes<T extends AnonymousRouter
       return next();
     }
   );
+
+  router.get(
+    '/saml-applications/:id/callback',
+    koaGuard({
+      params: z.object({ id: z.string() }),
+      query: samlApplicationSignInCallbackQueryParametersGuard,
+      status: [200, 400],
+    }),
+    async (ctx, next) => {
+      const {
+        params: { id },
+        query,
+      } = ctx.guard;
+
+      // Handle error in query parameters
+      if ('error' in query) {
+        throw new RequestError({
+          code: 'oidc.invalid_request',
+          message: query.error_description,
+        });
+      }
+
+      // Get application configuration
+      const {
+        secret,
+        oidcClientMetadata: { redirectUris },
+      } = await applications.findApplicationById(id);
+
+      assertThat(redirectUris[0], 'oidc.redirect_uri_not_set');
+
+      // TODO: should be able to handle `state` and code verifier etc.
+      const { code } = query;
+
+      // Handle OIDC callback and get user info
+      const userInfo = await handleOidcCallbackAndGetUserInfo(
+        code,
+        id,
+        secret,
+        redirectUris[0],
+        envSet.oidc.issuer
+      );
+
+      // Get SAML configuration
+      const { metadata } = await getSamlIdPMetadataByApplicationId(id);
+      const { privateKey } =
+        await samlApplicationSecrets.findActiveSamlApplicationSecretByApplicationId(id);
+      const { entityId, acsUrl } =
+        await samlApplicationConfigs.findSamlApplicationConfigByApplicationId(id);
```

Comment:
The `getSamlIdPMetadataByApplicationId` method internally calls both the `findActiveSamlApplicationSecretByApplicationId` and `findSamlApplicationConfigByApplicationId` queries already.

Reply:
Good call, this is refactored in another [PR#6887](https://github.com/logto-io/logto/pull/6887/files#diff-b01f544a43c34478115c852a6713775f661f159bbda5d6d59a01eaa677c1e97a).

Reply:
discussed offline, could leave a TODO here as well.